src/base/memory_monitor_sr.hpp

```diff
@@ -0,0 +1,273 @@
+    std::atomic<uint64_t> cur_stat;
+    std::atomic<uint64_t> peak_stat;
+    std::atomic<uint32_t> expand_count;
+  } MMM_MEM_STAT;
+
+  class mmm_subcomponent
+  {
+    public:
+      mmm_subcomponent (const char *name)
+      {
+	m_subcomp_name = new char[strlen (name) + 1];
+	strcpy (m_subcomp_name, name);
+      }
+
+      /* function */
+      const char *get_name ();
+      void add_cur_stat (uint64_t size);
+      void sub_cur_stat (uint64_t size);
+      const uint64_t get_cur_stat();
+
+      ~mmm_subcomponent ()
+      {
+	delete[] m_subcomp_name;
+      }
+
+    private:
+      /* private variable */
+      char *m_subcomp_name;
+      std::atomic<uint64_t> m_cur_stat;
+  };
+
+  class mmm_component
+  {
+    public:
+      mmm_component (const char *name)
+      {
+	m_comp_name = new char[strlen (name) + 1];
+	strcpy (m_comp_name, name);
+      }
+
+      /* function */
+      const char *get_name ();
+      void add_init_stat (int size);
+      void sub_init_stat (int size);
+      void add_cur_stat (int size);
+      void sub_cur_stat (int size);
+      void add_peak_stat (int size);
+      void sub_peak_stat (int size);
+      void add_expand_count ();
+      const MMM_MEM_STAT &get_stat();
+      const std::vector<mmm_subcomponent *> &get_subcomp_vec ();
+      int add_subcomponent (const char *name);
+
+      ~mmm_component ()
+      {
+	delete[] m_comp_name;
+      }
+    private:
+      /* private variable */
+      char *m_comp_name;
+      MMM_MEM_STAT m_stat;
+      std::vector<mmm_subcomponent *> m_subcomponent;
+  };
+
+  class mmm_module
+  {
+    public:
+      /* const */
+      /* max index of component or subcomponent is 0x0000FFFF
+       * if some stats have max index of component or subcomponent,
+       * we don't have to increase it */
+      static constexpr int MAX_COMP_IDX = 0x0000FFFF;
+
+    public:
+      mmm_module (const char *name, const MMM_COMP_INFO *info);
+      mmm_module () {} /* for dummy */
+
+      /* function */
+      virtual int aggregate_stats (MEMMON_MODULE_INFO *info);
+      const char *get_name ();
+      void add_init_stat (int size);
+      void sub_init_stat (int size);
+      void add_cur_stat (int size);
+      void sub_cur_stat (int size);
+      void add_peak_stat (int size);
+      void sub_peak_stat (int size);
+      void add_expand_count ();
+      const MMM_MEM_STAT &get_stat ();
+      int add_component (const char *name);
+      const std::vector<mmm_component *> &get_comp_vec ();
+      const std::vector<int> &get_comp_idx_map ();
+      void add_comp_idx_map (int compound);
+
+      static inline int MMM_MAKE_STAT_IDX_MAP (int comp_idx, int subcomp_idx)
+      {
+	return (comp_idx << 16 | subcomp_idx);
+      }
+
+      virtual ~mmm_module ()
+      {
+	if (m_module_name)
+	  {
+	    delete[] m_module_name;
+	  }
+      }
+    private:
+      /* private variable */
+      char *m_module_name;
+      MMM_MEM_STAT m_stat;
+      std::vector<mmm_component *> m_component;
```

Comment:
Why do you use `mmm_component *` instead of `mmm_component`?
If it is allocated every time add_component() is called, then I think it would be good to use `std::unique_ptr`.

Reply:
Change it to unique_ptr.